pkg/apis/resolution/v1alpha1/resolution_request_types.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2022 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ResolutionRequest is an object for requesting the content of
+// a Tekton resource like a pipeline.yaml.
+//
+// +genclient
+// +genreconciler
+type ResolutionRequest struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec holds the information for the request part of the resource request.
+	// +optional
+	Spec ResolutionRequestSpec `json:"spec,omitempty"`
+
+	// Status communicates the state of the request and, ultimately,
+	// the content of the resolved resource.
+	// +optional
+	Status ResolutionRequestStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ResolutionRequestList is a list of ResolutionRequests.
+type ResolutionRequestList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata"`
+	Items           []ResolutionRequest `json:"items"`
+}
+
+// ResolutionRequestSpec are all the fields in the spec of the
+// ResolutionRequest CRD.
+type ResolutionRequestSpec struct {
+	// Parameters are the runtime attributes passed to
+	// the resolver to help it figure out how to resolve the
+	// resource being requested. For example: repo URL, commit SHA,
+	// path to file, the kind of authentication to leverage, etc.
+	// +optional
+	Parameters map[string]string `json:"params,omitempty"`
```

Comment:
I was sharing the remote resource spec with a team and came across two different syntaxes. It seems we are implementing it using `params`:

```
taskRef:
    resolver: git
    resource:
    - name: url
      value: https://github.com/tekton/catalog.git
    - name: path
      value: /task/golang-fuzz/0.1/golang-fuzz.yaml
    - name: branch
      value: main
```

The [proposal](https://github.com/tektoncd/community/blob/main/teps/0060-remote-resource-resolution.md#proposal) in the Remote Resolution TEP has listed it as:

```
taskRef:
  resolver: git
  resource:
    repository_url: https://github.com/tektoncd/catalog.git
    branch: main
    path: /task/golang-fuzz/0.1/golang-fuzz.yaml
```

Do we need to update the proposal to reflect `params`? or is it documented somewhere else?

Reply:
Good question! I didn't do the initial implementation, but while I like the `repository_url` etc syntax better, the `- name: url` syntax is simpler to implement, since we don't need to use a `runtime.RawExtension` here for arbitrary YAML. 

So I'd say the proposal should be updated, yeah.

Reply:
@abayer to play devil's advocate, we should aim for the API to be the easiest to use (for end-user) than the easiest to implement 😝

Reply:
Fair point. =) I spent some time this weekend thinking about this - in practice, it probably is easier to go with a simple `map[string]string` than I'd been thinking, given that what we've got now is already just a slice of string name/value pairs anyway. Then, if we hit a situation down the road where we have a need for more complex resolver parameters, we could change that to something like `unstructured.Unstructured` without breaking API compatibility.

Reply:
...though it is worth mentioning that changing the syntax will break compatibility with the existing syntax, but that's alpha syntax and probably not at all widely used right now anyway.

Reply:
> ...though it is worth mentioning that changing the syntax will break compatibility with the existing syntax, but that's alpha syntax and probably not at all widely used right now anyway.

yeah that's the only reason that was "against" changing, but this is behind the alpha feature-flag for that sole reason, so if we properly communicate, I think that is fine.

Reply:
👍 Lemme do that!

Reply:
the reason why it was implemented as such was because of K8s convention: "Lists of named subobjects preferred over maps" - as discussed in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps and https://github.com/kubernetes/kubernetes/issues/2004

and we listed in https://github.com/tektoncd/community/blob/main/design-principles.md#api-conventions that Tekton API should comply with the K8s API conventions

Reply:
haha just saw the comments below

Reply:
...and yeah, I reverted back to the old form for `ResolutionRequest.Spec` so as not to break compatibility. See https://github.com/tektoncd/pipeline/pull/5200#issuecomment-1219463898 =)